integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ObservationFhirResourceProviderIntegrationTest.java

```diff
@@ -545,4 +547,483 @@ public void shouldReturnCountForObservationAsXml() throws Exception {
+		assertThat(entries, everyItem(hasResource(
+		    hasProperty("code", hasProperty("coding", everyItem(hasProperty("display", equalTo("Respiratory rate"))))))));
+
+		assertThat(checkDistinctObsDateTimeInEachCodeGroupIsSortedAndWithinMax(entries, 2), is(true));
+	}
+	
+	@Test
+	public void shouldReturnLastnObservationsWhenCodeIsMissingAsXml() throws Exception {
+		MockHttpServletResponse response = get(
+		    "Observation/$lastn?max=2&subject=" + OBS_PATIENT_UUID + "&category=laboratory").accept(FhirMediaTypes.XML).go();
+		
+		assertThat(response, isOk());
+		assertThat(response.getContentType(), is(FhirMediaTypes.XML.toString()));
+		assertThat(response.getContentAsString(), notNullValue());
+		
+		Bundle results = readBundleResponse(response);
+		
+		assertThat(results, notNullValue());
+		assertThat(results.getType(), equalTo(Bundle.BundleType.COLLECTION));
+		assertThat(results.hasEntry(), is(true));
+		assertThat(results, hasProperty("total", equalTo(17)));
+		
+		List<Bundle.BundleEntryComponent> entries = results.getEntry();
+		
+		assertThat(entries, everyItem(hasProperty("fullUrl", startsWith("http://localhost/ws/fhir2/R4/Observation/"))));
+		assertThat(entries, everyItem(hasResource(instanceOf(Observation.class))));
+		assertThat(entries, everyItem(hasResource(validResource())));
+		assertThat(entries,
+		    everyItem(hasResource(hasProperty("subject", hasProperty("reference", endsWith(OBS_PATIENT_UUID))))));
+		assertThat(entries, everyItem(hasResource(hasProperty("category",
+		    everyItem(hasProperty("coding", everyItem(hasProperty("code", equalTo("laboratory")))))))));
+
+		assertThat(checkDistinctObsDateTimeInEachCodeGroupIsSortedAndWithinMax(entries, 2), is(true));
+	}
+	
+	@Test
+	public void shouldReturnLastnObservationsWhenNoParamterIsGivenAsXml() throws Exception {
+		MockHttpServletResponse response = get("Observation/$lastn?").accept(FhirMediaTypes.XML).go();
+		
+		assertThat(response, isOk());
+		assertThat(response.getContentType(), is(FhirMediaTypes.XML.toString()));
+		assertThat(response.getContentAsString(), notNullValue());
+		
+		Bundle results = readBundleResponse(response);
+		
+		assertThat(results, notNullValue());
+		assertThat(results.getType(), equalTo(Bundle.BundleType.COLLECTION));
+		assertThat(results.hasEntry(), is(true));
+		assertThat(results, hasProperty("total", equalTo(14)));
+		
+		List<Bundle.BundleEntryComponent> entries = results.getEntry();
+		
+		assertThat(entries, everyItem(hasProperty("fullUrl", startsWith("http://localhost/ws/fhir2/R4/Observation/"))));
+		assertThat(entries, everyItem(hasResource(instanceOf(Observation.class))));
+		assertThat(entries, everyItem(hasResource(validResource())));
+
+		assertThat(checkDistinctObsDateTimeInEachCodeGroupIsSortedAndWithinMax(entries, 1), is(true));
+	}
+	
+	private boolean checkDistinctObsDateTimeInEachCodeGroupIsSortedAndWithinMax(List<Bundle.BundleEntryComponent> entries, int max) {
```

Comment:
This should probably be implemented as a custom matcher rather than a long test that only returns a boolean value. The advantage is that a custom matcher allows us to actually describe the error that caused the test to fail with as much detail as possible. See, e.g., [this article](https://www.baeldung.com/hamcrest-custom-matchers) or [some of the custom matchers in the BaseFhirIntegrationTest class](https://github.com/openmrs/openmrs-module-fhir2/blob/c076dcbe1c89ce68098788333eca3e42e6a89f0d/integration-tests/src/test/java/org/openmrs/module/fhir2/BaseFhirIntegrationTest.java#L275) for what this might look like.

Reply:
I'll try implementing a custom matcher for this.